File: hw/motor.py
"""
High-level RS03 motor control API.

Wraps robstride.Client + RobStrideBus into a clean interface
for controlling RS03 actuators on the robot arm.

Usage:
    from hw import Motor

    m = Motor("COM5", can_id=127)
    m.set_zero()
    m.move_to(1.57)       # go to 90 degrees
    m.spin(5.0)           # 5 rad/s
    m.stop()
    m.close()
"""

import time
import math
from dataclasses import dataclass
from typing import Optional

from hw.robstride_bus import RobStrideBus
from robstride import Client, RunMode
from robstride.client import FeedbackResp, MotorMode


# RS03 limits (from RobStride documentation)
RS03_MAX_TORQUE = 60.0    # N*m peak
RS03_MAX_SPEED = 50.0     # rad/s
RS03_MAX_CURRENT = 30.0   # A


@dataclass
class MotorState:
    angle: float       # radians
    velocity: float    # rad/s
    torque: float      # N*m
    temperature: float # Celsius
    mode: str
    errors: list


class Motor:
    """High-level controller for a single RS03 motor."""

    def __init__(
        self,
        port: str,
        can_id: int,
        host_id: int = 0xAA,
        baud: int = 921600,
        bus: Optional[RobStrideBus] = None,
    ):
        self.can_id = can_id
        self._owns_bus = bus is None
        self._bus = bus or RobStrideBus(port, ttyBaudrate=baud)
        self._client = Client(self._bus, host_can_id=host_id)
        self._enabled = False

    # ── Lifecycle ────────────────────────────────────────────

    def enable(self) -> MotorState:
        """Enable the motor. Must be called before any motion commands."""
        fb = self._client.enable(self.can_id)
        self._enabled = True
        return self._to_state(fb)

    def disable(self) -> MotorState:
        """Disable the motor (coast to stop)."""
        fb = self._client.disable(self.can_id)
        self._enabled = False
        return self._to_state(fb)

    def stop(self) -> MotorState:
        """Alias for disable."""
        return self.disable()

    def close(self):
        """Disable motor and release the serial port."""
        if self._enabled:
            try:
                self.disable()
            except Exception:
                pass
        if self._owns_bus:
            self._bus.shutdown()

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()

    # ── Configuration ────────────────────────────────────────

    def set_zero(self):
        """Set the current position as mechanical zero."""
        self._client.bus.send(self._client._rs_msg(
            type('', (), {'value': 6})(),  # ZeroPos = 6
            self._client.host_can_id,
            self.can_id,
            [0, 0, 0, 0, 0, 0, 0, 0],
        ))
        self._client._recv()

    def set_run_mode(self, mode: RunMode):
        """Set the control mode: Position, Speed, Current, or Operation (MIT)."""
        self._client.write_param(self.can_id, 'run_mode', mode)

    def set_position_gain(self, kp: float):
        self._client.write_param(self.can_id, 'loc_kp', kp)

    def set_speed_gain(self, kp: float, ki: float = 0.02):
        self._client.write_param(self.can_id, 'spd_kp', kp)
        self._client.write_param(self.can_id, 'spd_ki', ki)

    def set_speed_limit(self, limit: float):
        """Set max speed in rad/s (applies in position mode)."""
        self._client.write_param(self.can_id, 'limit_spd', min(limit, RS03_MAX_SPEED))

    def set_torque_limit(self, limit: float):
        """Set max torque in N*m."""
        self._client.write_param(self.can_id, 'limit_torque', min(limit, RS03_MAX_TORQUE))

    # ── Motion commands ──────────────────────────────────────

    def move_to(self, position_rad: float, speed_limit: Optional[float] = None):
        """Move to an absolute position in radians (position mode).

        The motor must be enabled first. This sets position mode
        and sends the target. Returns immediately; the motor
        moves asynchronously.
        """
        self._ensure_enabled()
        self.set_run_mode(RunMode.Position)
        if speed_limit is not None:
            self.set_speed_limit(speed_limit)
        self._client.write_param(self.can_id, 'loc_ref', position_rad)

    def move_to_deg(self, degrees: float, speed_limit: Optional[float] = None):
        """Move to an absolute position in degrees."""
        self.move_to(math.radians(degrees), speed_limit)

    def spin(self, velocity_rad_s: float):
        """Spin at a constant velocity in rad/s (speed mode)."""
        self._ensure_enabled()
        self.set_run_mode(RunMode.Speed)
        self._client.write_param(self.can_id, 'spd_ref', velocity_rad_s)

    def set_torque(self, torque_nm: float):
        """Apply a raw torque in N*m (current/torque mode)."""
        self._ensure_enabled()
        self.set_run_mode(RunMode.Current)
        self._client.write_param(self.can_id, 'iq_ref', torque_nm)

    # ── Telemetry ────────────────────────────────────────────

    def read_state(self) -> MotorState:
        """Read current motor state (position, velocity, torque, temp)."""
        fb = self._client.enable(self.can_id)
        return self._to_state(fb)

    def read_param(self, name: str):
        """Read a raw parameter by name."""
        return self._client.read_param(self.can_id, name)

    def read_voltage(self) -> float:
        return self._client.read_param(self.can_id, 'vbus')

    def read_position(self) -> float:
        return self._client.read_param(self.can_id, 'mechpos')

    def read_velocity(self) -> float:
        return self._client.read_param(self.can_id, 'mechvel')

    # ── Helpers ──────────────────────────────────────────────

    def wait_until_at(self, target_rad: float, tolerance: float = 0.05, timeout: float = 10.0):
        """Block until motor is within tolerance of target position."""
        start = time.time()
        while time.time() - start < timeout:
            pos = self.read_position()
            if abs(pos - target_rad) < tolerance:
                return True
            time.sleep(0.05)
        return False

    def _ensure_enabled(self):
        if not self._enabled:
            self.enable()

    @staticmethod
    def _to_state(fb: FeedbackResp) -> MotorState:
        return MotorState(
            angle=fb.angle,
            velocity=fb.velocity,
            torque=fb.torque,
            temperature=fb.temp,
            mode=fb.mode.name,
            errors=[e.name for e in fb.errors],
        )

File: src/motor.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use anyhow::Result;
use robstride::{
    Command, CommandData, EnableCommand, FeedbackFrame, MotorMode, Protocol, ReadCommand,
    SetZeroCommand, StopCommand, TransportType, WriteCommand,
};
use robstride::actuator::TypedFeedbackData;
use robstride::robstride03::{RobStride03Feedback, RobStride03Parameter};
use robstride::ActuatorParameter;
use tokio::sync::Mutex;

const HOST_ID: u8 = 0xAA;

pub struct MotorState {
    pub angle_rad: f32,
    pub velocity_rads: f32,
    pub torque_nm: f32,
    pub temperature_c: f32,
    pub mode: MotorMode,
    pub faults: Vec<&'static str>,
}

pub struct Motor {
    protocol: Arc<Mutex<Protocol>>,
    pub can_id: u8,
    host_id: u8,
    enabled: bool,
    pub debug: bool,
}

impl Motor {
    pub fn new(protocol: Arc<Mutex<Protocol>>, can_id: u8) -> Self {
        Self {
            protocol,
            can_id,
            host_id: HOST_ID,
            enabled: false,
            debug: false,
        }
    }

    pub fn with_host_id(mut self, host_id: u8) -> Self {
        self.host_id = host_id;
        self
    }

    // -- Lifecycle --

    pub async fn enable(&mut self) -> Result<MotorState> {
        let cmd = EnableCommand {
            host_id: self.host_id,
        };
        let (id, data) = cmd.to_can_packet(self.can_id);
        let fb = self.send_and_recv(id, &data).await?;
        self.enabled = true;
        Ok(Self::parse_feedback(fb))
    }

    pub async fn disable(&mut self) -> Result<MotorState> {
        let cmd = StopCommand {
            host_id: self.host_id,
            clear_fault: false,
        };
        let (id, data) = cmd.to_can_packet(self.can_id);
        let fb = self.send_and_recv(id, &data).await?;
        self.enabled = false;
        Ok(Self::parse_feedback(fb))
    }

    pub async fn set_zero(&mut self) -> Result<()> {
        let cmd = SetZeroCommand {
            host_id: self.host_id,
        };
        let (id, data) = cmd.to_can_packet(self.can_id);
        self.send_and_recv(id, &data).await?;
        Ok(())
    }

    // -- Configuration --

    pub async fn set_run_mode(&mut self, mode: u8) -> Result<()> {
        self.write_param(RobStride03Parameter::RunMode, mode as f32).await
    }

    pub async fn set_speed_limit(&mut self, limit_rads: f32) -> Result<()> {
        self.write_param(RobStride03Parameter::LimitSpd, limit_rads.min(20.0)).await
    }

    pub async fn set_torque_limit(&mut self, limit_nm: f32) -> Result<()> {
        self.write_param(RobStride03Parameter::LimitTorque, limit_nm.min(60.0)).await
    }

    pub async fn set_position_gain(&mut self, kp: f32) -> Result<()> {
        self.write_param(RobStride03Parameter::LocKp, kp).await
    }

    pub async fn set_speed_gain(&mut self, kp: f32, ki: f32) -> Result<()> {
        self.write_param(RobStride03Parameter::SpdKp, kp).await?;
        self.write_param(RobStride03Parameter::SpdKi, ki).await
    }

    // -- Motion --

    pub async fn move_to(&mut self, position_rad: f32, speed_limit: Option<f32>) -> Result<()> {
        self.ensure_enabled().await?;
        self.set_run_mode(1).await?;
        if let Some(limit) = speed_limit {
            self.set_speed_limit(limit).await?;
        }
        self.write_param(RobStride03Parameter::Ref, position_rad).await
    }

    pub async fn move_to_deg(&mut self, degrees: f32, speed_limit: Option<f32>) -> Result<()> {
        self.move_to(degrees.to_radians(), speed_limit).await
    }

    pub async fn spin(&mut self, velocity_rads: f32) -> Result<()> {
        self.ensure_enabled().await?;
        self.set_run_mode(2).await?;
        self.write_param(RobStride03Parameter::SpdRef, velocity_rads).await
    }

    pub async fn set_torque(&mut self, torque_nm: f32) -> Result<()> {
        self.ensure_enabled().await?;
        self.set_run_mode(3).await?;
        self.write_param(RobStride03Parameter::IqRef, torque_nm).await
    }

    // -- Telemetry --

    pub async fn read_state(&mut self) -> Result<MotorState> {
        let cmd = EnableCommand {
            host_id: self.host_id,
        };
        let (id, data) = cmd.to_can_packet(self.can_id);
        let fb = self.send_and_recv(id, &data).await?;
        Ok(Self::parse_feedback(fb))
    }

    pub async fn read_param(&mut self, param: RobStride03Parameter) -> Result<f32> {
        let meta = param.metadata();
        let cmd = ReadCommand {
            host_id: self.host_id,
            parameter_index: meta.index,
            data: 0,
            read_status: false,
        };
        let (id, data) = cmd.to_can_packet(self.can_id);
        if self.debug {
            eprintln!("  TX  id={:08X} data={:02X?}  (read param 0x{:04X} '{}')",
                id, data, meta.index, meta.name);
        }
        let mut proto = self.protocol.lock().await;
        proto.send(id, &data).await
            .map_err(|e| anyhow::anyhow!("{:#}", e))?;
        let (resp_id, resp_data) = proto.recv().await
            .map_err(|e| anyhow::anyhow!("{:#}", e))?;
        drop(proto);

        if self.debug {
            eprintln!("  RX  id={:08X} data={:02X?}", resp_id, &resp_data);
        }

        let resp_cmd = Command::from_can_packet(resp_id, resp_data);
        let read_resp = ReadCommand::from_command(resp_cmd);
        Ok(read_resp.data_as_f32())
    }

    pub async fn read_position(&mut self) -> Result<f32> {
        self.read_param(RobStride03Parameter::MechPos).await
    }

    pub async fn read_velocity(&mut self) -> Result<f32> {
        self.read_param(RobStride03Parameter::MechVel).await
    }

    pub async fn read_voltage(&mut self) -> Result<f32> {
        self.read_param(RobStride03Parameter::VBus).await
    }

    // -- Helpers --

    pub async fn wait_until_at(
        &mut self,
        target_rad: f32,
        tolerance: f32,
        timeout: Duration,
    ) -> Result<bool> {
        let start = Instant::now();
        while start.elapsed() < timeout {
            let pos = self.read_position().await?;
            if (pos - target_rad).abs() < tolerance {
                return Ok(true);
            }
            tokio::time::sleep(Duration::from_millis(50)).await;
        }
        Ok(false)
    }

    async fn ensure_enabled(&mut self) -> Result<()> {
        if !self.enabled {
            self.enable().await?;
        }
        Ok(())
    }

    async fn write_param(&mut self, param: RobStride03Parameter, value: f32) -> Result<()> {
        let meta = param.metadata();
        if self.debug {
            eprintln!("  WRITE param 0x{:04X} '{}' = {} (type={:?})",
                meta.index, meta.name, value, meta.param_type);
        }
        let cmd = WriteCommand {
            host_id: self.host_id,
            parameter_index: meta.index,
            data: value,
        };
        let (id, data) = cmd.to_can_packet(self.can_id);
        self.send_and_recv(id, &data).await?;
        Ok(())
    }

    async fn send_and_recv(&self, id: u32, data: &[u8]) -> Result<FeedbackFrame> {
        if self.debug {
            let comm_type = (id >> 24) & 0x1F;
            let data_2 = (id >> 8) & 0xFFFF;
            let target = id & 0x7F;
            eprintln!("  TX  id={:08X} [comm={} d2={:04X} tgt={}] data={:02X?}",
                id, comm_type, data_2, target, data);
        }
        let mut proto = self.protocol.lock().await;
        proto.send(id, data).await
            .map_err(|e| anyhow::anyhow!("{:#}", e))?;
        let (resp_id, resp_data) = proto.recv().await
            .map_err(|e| anyhow::anyhow!("{:#}", e))?;
        drop(proto);

        if self.debug {
            let comm_type = (resp_id >> 24) & 0x1F;
            let data_2 = (resp_id >> 8) & 0xFFFF;
            let target = resp_id & 0x7F;
            eprintln!("  RX  id={:08X} [comm={} d2={:04X} tgt={}] data={:02X?}",
                resp_id, comm_type, data_2, target, &resp_data);
        }

        let cmd = Command::from_can_packet(resp_id, resp_data);
        let fb = FeedbackFrame::from_command(cmd);
        Ok(fb)
    }

    fn parse_feedback(fb: FeedbackFrame) -> MotorState {
        let typed = RobStride03Feedback::from_feedback_frame(fb.clone());

        let mut faults = Vec::new();
        if fb.fault_undervoltage { faults.push("undervoltage"); }
        if fb.fault_overcurrent { faults.push("overcurrent"); }
        if fb.fault_over_temperature { faults.push("over_temperature"); }
        if fb.fault_magnetic_encoding { faults.push("magnetic_encoding"); }
        if fb.fault_hall_encoding { faults.push("hall_encoding"); }
        if fb.fault_uncalibrated { faults.push("uncalibrated"); }

        MotorState {
            angle_rad: typed.angle_rad(),
            velocity_rads: typed.velocity_rads(),
            torque_nm: typed.torque_nm(),
            temperature_c: fb.temperature,
            mode: fb.mode,
            faults,
        }
    }
}

pub async fn create_ch341_protocol(port: &str) -> Result<Arc<Mutex<Protocol>>> {
    let transport = robstride::CH341Transport::new(port.to_string())
        .await
        .map_err(|e| anyhow::anyhow!("Failed to open CH341 on {}: {:#}", port, e))?;
    let transport = TransportType::CH341(transport);
    let callback = Arc::new(|_id: u32, _data: Vec<u8>| {});
    let protocol = Protocol::new(transport, callback);
    Ok(Arc::new(Mutex::new(protocol)))
}
